### ISSUE TITLE
[Feature] `slug` field type

### ISSUE THREAD

**Issue by tabacitu** [Laravel-Backpack/ideas#127]
Eg:
- Title [ _____________ ]
- Slug   [ ______________]

I think it should be pretty easy to create a field that will take the contents of `title` as you type them, slug them, and put them into the `slug` field. We've probably dismissed this until now, thinking that it'll be difficult to re-create _the same REGEX on the front-end_ as will be done by the PHP on the back-end, but... it doesn't _really_ have to be perfectly the same... it just needs to be a valid slug, after all. This was my 🤯 moment today 😀

Another reason we don't have this until now is that... I always thought this field would also need real-time validation. To check if that slug is available. But you know what? No... it doesn't _need_ that. It would be nice (sure!) but... it doesn't need it.

I've been overthinking this field, I think. And because of that, we don't have a solution right now. We could easily create a solution for the 90% use case... so... why don't we?

So... I think we should do this. I think we should have a `slug` field type. We said we'd do documentation-driven-development, so here it is:

```php
CRUD::addField([
    'name' => 'slug',
    'type' => 'slug',
    'label' => 'Slug',
    'hint' => 'Unique identifying segment for this item, used in URLs.',
    'targetInputName' => 'title',
    'slugWhenNotEmpty' => false, // change the existing slug, if "title" changes?
]);
```

**Comment by o15a3d4l11s2** [comment]
Here is an old filed that I have used. Please note that I have created it for Backpack 4 long time ago.

As a comparison to your documentation I needed the field to generate a slug only when not manually changed. Even is value is manually deleted in my case I do not re-generate it. But in case of editing I store the original slug so that I can restore it. Also my implementation uses cviebrock/eloquent-sluggable.

`slug.blade.php
```
<!-- text input -->
@include('crud::fields.inc.wrapper_start')
<label>{!! $field['label'] !!}</label>
@include('crud::fields.inc.translatable_icon')

@if(isset($field['prefix']) || isset($field['suffix']))
    <div class="input-group"> @endif
        @if(isset($field['prefix']))
            <div class="input-group-prepend"><span class="input-group-text">{!! $field['prefix'] !!}</span></div> @endif
        <input
            type="text"
            data-init-function="bpFieldInitSlugElement"
            name="{{ $field['name'] }}"
            value="{{ old(square_brackets_to_dots($field['name'])) ?? $field['value'] ?? $field['default'] ?? '' }}"
            data-source-field-name="{{ $field['source_field_name'] ?? 'name'  }}"
            @include('crud::fields.inc.attributes')
        >
        @if(isset($field['suffix']))
            <div class="input-group-append"><span class="input-group-text">{!! $field['suffix'] !!}</span></div> @endif
        @if(isset($field['prefix']) || isset($field['suffix'])) </div> @endif

{{-- HINT --}}
@if (isset($field['hint']))
    <p class="help-block">{!! $field['hint'] !!}</p>
    @endif
    </div>

    {{-- ########################################## --}}
    {{-- Extra CSS and JS for this particular field --}}
    {{-- If a field type is shown multiple times on a form, the CSS and JS will only be loaded once --}}
    @if ($crud->fieldTypeNotLoaded($field))
        @php
            $crud->markFieldTypeAsLoaded($field)
        @endphp

        {{-- FIELD JS - will be loaded in the after_scripts section --}}
        @push('crud_fields_scripts')
            {{-- YOUR JS HERE --}}
            <script>
                function bpFieldInitSlugElement(element) {
                    var $sourceFieldName = element.data('source-field-name');
                    var manuallyChanged = false;
                    var originalValue = element.val();
                    jQuery(document).ready(function ($) {
                        $('input[name="' + $sourceFieldName + '"]').on('focusout', function (e) {
                            if (element.val()) {
                                return;
                            }

                            if (manuallyChanged && !element.val()) {
                                element.val(originalValue);
                                return
                            }

                            $.get('{{ url($crud->route . '/slug') }}',
                                {'name': $(this).val()},
                                function (data) {
                                    element.val(data.slug);
                                    manuallyChanged = false;
                                }
                            );
                        })

                        element.on('input', function (e) {
                            manuallyChanged = true;
                        })
                    });
                }
            </script>
        @endpush
    @endif
    {{-- End of Extra CSS and JS --}}
    {{-- ########################################## --}}
```

and here is `Sluggable` trait:
```
trait Sluggable
{
    protected function setupSlugRoutes($segment, $routeName, $controller)
    {
        Route::get($segment . '/slug', [
            'as' => $routeName . '.generateSlug',
            'uses' => $controller . '@generateSlug',
            'operation' => 'generateSlug',
        ]);
    }

    protected function generateSlug()
    {
        $slugSource = $this->crud->getRequest()->input('name');
        $slug = null;
        if ($slugSource) {
            $slug = SlugService::createSlug(CRUD::getModel(), 'slug', $slugSource);
        }

        return Response::json(['slug' => $slug]);
    }
}
```

In summary I have few questions:
1. What is your preferred way of generating the slug? Through JS like your demo implementation of the new [JS library](https://github.com/Laravel-Backpack/CRUD/pull/4312)?
2. I agree that it is not absolutely required to have a realtime validation of the slug. But in cases where admin panel allows duplication of records this will be very handy.
3. What is your feeling about using cviebrock/eloquent-sluggable for the slugging or it is too much and too binding?

**Comment by tabacitu** [comment]
Hey @o15a3d4l11s2 ,

Thanks a lot for sharing this!

1. In this case, for a `slug` field, yes, I think the best way to generate the slug would be using JS, on the front-end. Easy-peasy.
2. True
3. I like it. It's a more strict implementation, I agree 😀 I use cviebrock/eloquent-sluggable on the back-end too... all the time. But for a `slug` field... seems a bit overkill. I mean, there's a tradeoff to be made here:
- (A) Best DX - do you make the field easy-to-use by the dev, by not needing a Route, Operation, packages installed? OR
- (B) Best UX - do you make the field easy-to-use by the user, by knowing when a route is taken?

In this case, I'd opt for the best DX, because I think it also provides a pretty good UX. The only use case where it fails a bit is when the route is taken, in which case validation will show that... so it's definitely not a bad UX either.

To sum up... I'd go with JS, to be honest, just seems like the less-complicated way to do it, for most use cases.
That's my 2 cents.

**Comment by tabacitu** [comment]
![animation-news-3o6ZtbRh0xqwWzO8PC](https://user-images.githubusercontent.com/1032474/171995015-eb2fe9ae-b167-42af-8f85-41b456b9902b.gif)

I've just created a 99% ready field for this, in PRO. It's so so simple and efficient!
We'll probably launch it next week, after it gets reviewed 🎉 [Subscribe to the blog article digest](https://backpackforlaravel.com/user/newsletters) to find out when it's out, or follow us on Twitter/Fb/smth.

But yeah... just look at it, so simple and beautiful 😀

![slug_field](https://user-images.githubusercontent.com/1032474/171995040-d8dbbfa2-b798-4452-9ffb-0a48665a0d31.gif)

I'll close this issue since we have a PR for it here - https://github.com/DigitallyHappy/backpack-pro/pull/46

Cheers!
